
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Welcome to PRSKB’s CLI documentation! &#8212; Polygenic Risk Score Knowledge Base  documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="welcome-to-prskb-s-cli-documentation">
<h1>Welcome to PRSKB’s CLI documentation!<a class="headerlink" href="#welcome-to-prskb-s-cli-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<p>The PRSKB’s CLI tool is an extension of the PRSKB web application. It is designed to give more flexability and capability in calculating polygenic risk scores for large datasets. Features include:</p>
<ul class="simple">
<li><p>Searching the database for studies and traits</p></li>
<li><p>Printing out available ethnicities to filter by</p></li>
<li><p>Learning about required and optional parameters for performing calculations</p></li>
<li><p>Calculating polygenic risk scores</p></li>
</ul>
<div class="section" id="quick-start">
<h2>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h2>
<p>To download the PRSKB CLI tool, head over to the <a class="reference external" href="https://prs.byu.edu/cli_download.html">PRSKB website download page</a> or download the whole repository directly from <a class="reference external" href="https://github.com/kauwelab/PolyRiskScore">GitHub</a>.</p>
<p>Given the required parameters, the tool will calculate risk scores for each individual sample for each study and trait combination in the database.</p>
<p><em>NOTE: You MUST have bash and jq for bash, and python3 and the PyVCF, filelock, and requests Python modules installed.</em></p>
<div class="section" id="required-parameters-example">
<h3>Required Parameters Example<a class="headerlink" href="#required-parameters-example" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.05 -r hg19 -p EUR
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="features">
<h1>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h1>
<p>The CLI polygenic risk score calculator can be run directly from the command-line or through the CLI’s interactive menu. The menu includes the options to search the database and to learn more about the parameters involved in risk score calculations.
To access the menu, run the script without parameters. You will be given the option to see the usage statement or start the interactive menu.</p>
<img alt="CLI tool run without parameters" src="_images/RunCLIwithoutParams.png" />
<img alt="CLI menu screen shot" src="_images/RunMenu.png" />
<div class="section" id="search-the-database">
<h2>Search the Database<a class="headerlink" href="#search-the-database" title="Permalink to this headline">¶</a></h2>
<p>Through the interactive menu, users can search the database for traits or studies. This can be helpful when filtering studies to run the calculator on.</p>
<p>After selecting the menu option to search for studies or traits, you will be prompted to specify if you are searching for studies or traits. You can then enter the search term you wish to use.</p>
<img alt="Screenshot example for searching traits about teeth" src="_images/SearchTraits.png" />
<p>When the search is complete, the results will be printed to the console, and you will be returned to the menu.</p>
</div>
<div class="section" id="display-ethnicities">
<h2>Display Ethnicities<a class="headerlink" href="#display-ethnicities" title="Permalink to this headline">¶</a></h2>
<p>The menu has an option for displaying the available ethnicities to filter studies by.</p>
<img alt="Screenshot showing displayed ethnicities" src="_images/displayEthnicities.png" />
</div>
<div class="section" id="learn-about-parameters">
<h2>Learn about Parameters<a class="headerlink" href="#learn-about-parameters" title="Permalink to this headline">¶</a></h2>
<p>The menu also has an option for learning about the parameters involved in filtering studies and calculating scores.</p>
<img alt="Screen shot of Parameters menu" src="_images/LearnAboutParams.png" />
<div class="section" id="required-parameters">
<h3>Required Parameters<a class="headerlink" href="#required-parameters" title="Permalink to this headline">¶</a></h3>
<div class="section" id="file-path-f">
<h4>File Path (-f)<a class="headerlink" href="#file-path-f" title="Permalink to this headline">¶</a></h4>
<p>The path to the input file. Either a vcf or a txt with lines formatted as rsID:allele1,allele2</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-f path/to/file/samples.vcf
</pre></div>
</div>
</div>
<div class="section" id="output-file-o">
<h4>Output File (-o)<a class="headerlink" href="#output-file-o" title="Permalink to this headline">¶</a></h4>
<p>The path to output file. Can be either a tsv or a json.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-o path/to/file/output.tsv
</pre></div>
</div>
</div>
<div class="section" id="p-value-cutoff-c">
<h4>P-value Cutoff (-c)<a class="headerlink" href="#p-value-cutoff-c" title="Permalink to this headline">¶</a></h4>
<p>Creates a threshold for included snps.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-c <span class="m">0</span>.0005
</pre></div>
</div>
</div>
<div class="section" id="reference-genome-r">
<h4>Reference Genome (-r)<a class="headerlink" href="#reference-genome-r" title="Permalink to this headline">¶</a></h4>
<p>The reference genome of samples in the input file. Available reference genomes: hg17, hg18, hg19, or hg38</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-r hg19
</pre></div>
</div>
</div>
<div class="section" id="super-population-p">
<h4>Super Population (-p)<a class="headerlink" href="#super-population-p" title="Permalink to this headline">¶</a></h4>
<p>The super population of samples in the input file. These are the five super populations from the 1000 genomes project</p>
<ul class="simple">
<li><p>AFR - African</p></li>
<li><p>AMR - American</p></li>
<li><p>EAS - East Asian</p></li>
<li><p>EUR - European</p></li>
<li><p>SAS - South Asian</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-p EUR
</pre></div>
</div>
</div>
</div>
<div class="section" id="optional-filtering-parameters">
<h3>Optional Filtering Parameters<a class="headerlink" href="#optional-filtering-parameters" title="Permalink to this headline">¶</a></h3>
<p>To run the PRSKB CLI calculator on a more specific set of studies, optional filtering parameters can be added. When specifying more than one, add the identifying flag in front of every filter specified.</p>
<div class="section" id="traits-t">
<h4>Traits (-t)<a class="headerlink" href="#traits-t" title="Permalink to this headline">¶</a></h4>
<p>Specific traits to run the calculator on.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-t <span class="s2">&quot;Alzheimer&#39;s Disease&quot;</span> -t acne
</pre></div>
</div>
</div>
<div class="section" id="study-types-k">
<h4>Study Types (-k)<a class="headerlink" href="#study-types-k" title="Permalink to this headline">¶</a></h4>
<p>Specifies what kinds of studies to run analyses on.</p>
<ul class="simple">
<li><p>HI (high impact) - studies that have the highest altmetric score within their trait group</p></li>
<li><p>LC (largest cohort) - studies that have the largest sample size (initial and replication combined)</p></li>
<li><p>O (other) - all other studies that do not fall in the first two categories</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-k HI -k LC
</pre></div>
</div>
</div>
<div class="section" id="study-ids-i">
<h4>Study IDs (-i)<a class="headerlink" href="#study-ids-i" title="Permalink to this headline">¶</a></h4>
<p>Specifies the studyID(s) of studies to run the calculator on.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-i GCST004410
</pre></div>
</div>
</div>
<div class="section" id="ethnicity-e">
<h4>Ethnicity (-e)<a class="headerlink" href="#ethnicity-e" title="Permalink to this headline">¶</a></h4>
<p>Filters studies by the ethnicity of their sample population.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-e European -e <span class="s2">&quot;East Asian&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="additonal-optional-parameters">
<h3>Additonal Optional Parameters<a class="headerlink" href="#additonal-optional-parameters" title="Permalink to this headline">¶</a></h3>
<p>These additional parameters manage calculation details and may only be specified once per calculation.</p>
<div class="section" id="verbose-v">
<h4>Verbose (-v)<a class="headerlink" href="#verbose-v" title="Permalink to this headline">¶</a></h4>
<p>Creates a more detailed TSV result file.
The verbose output file will include the following for each corresponding sample, study, and trait combination:</p>
<ul class="simple">
<li><p>reported trait</p></li>
<li><p>trait</p></li>
<li><p>polygenic risk score</p></li>
<li><p>protective variants</p></li>
<li><p>risk variants</p></li>
<li><p>variants that are present but do not include the risk allele</p></li>
<li><p>variants that are in high linkage disequilibrium whose odds ratios are not included in the calculations</p></li>
</ul>
<p>If the output file is in TSV format and this parameter is not included, the default TSV result
file will include the study ID and the corresponding polygenic risk scores for each sample.
If the output file is in JSON format, the results will, by default, be in verbose format.</p>
<p><em>NOTE: There is no condensed version of JSON output.</em></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-v
</pre></div>
</div>
</div>
<div class="section" id="default-sex-g">
<h4>Default Sex (-g)<a class="headerlink" href="#default-sex-g" title="Permalink to this headline">¶</a></h4>
<p>Though a rare occurence, some studies have duplicates of the same snp that differ by which
biological sex the p-value is associated with. You can indicate which sex you would like snps
to select when both options (M/F) are present. The system default is Female.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-g f
</pre></div>
</div>
</div>
<div class="section" id="step-number-s">
<h4>Step Number (-s)<a class="headerlink" href="#step-number-s" title="Permalink to this headline">¶</a></h4>
<p>Breaks running the calculator into steps. Make sure when you are running the CLI in steps that the only parameter that changes between the two steps is the step parameter.</p>
<p>Step 1 deals with downloading data from the server for use in calculations. This step requires internet access.</p>
<p>Step 2 deals with the actual calculation of polygenic risk scores. This step does not require internet access but does need to have access to the files that were downloaded in step 1.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-s <span class="m">1</span>
</pre></div>
</div>
</div>
<div class="section" id="number-of-processes-n">
<h4>Number Of Processes (-n)<a class="headerlink" href="#number-of-processes-n" title="Permalink to this headline">¶</a></h4>
<p>This parameter determines the number of subprocesses used by the Python multiprocessing module during the calculations. If no value is given by the user, all available cores will be utilized.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-n <span class="m">4</span>
</pre></div>
</div>
</div>
<div class="section" id="omit-unused-studies-file-m">
<h4>Omit Unused Studies file (-m)<a class="headerlink" href="#omit-unused-studies-file-m" title="Permalink to this headline">¶</a></h4>
<p>This flag will prevent the creation of an additional output file that lists the studies that no risk scores could be calculated for.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-m
</pre></div>
</div>
</div>
<div class="section" id="user-gwas-upload-file-u">
<h4>User GWAS upload file (-u)<a class="headerlink" href="#user-gwas-upload-file-u" title="Permalink to this headline">¶</a></h4>
<p>This parameter allows the user to upload a GWAS summary statistics file to be used in polygenic risk score calculations instead of GWAS Catalog data stored in our database. The file must be tab separated, use a .tsv or .txt extension (or be a zipped file with one of those extensions), and have the correct columns in order for calculations to occur. (see Uploading GWAS Summary Statistics for more directions on uploading GWAS data)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-u path/to/file/GWASsummaryStatistics.tsv
</pre></div>
</div>
</div>
<div class="section" id="gwas-reference-genome-a">
<h4>GWAS reference genome (-a)<a class="headerlink" href="#gwas-reference-genome-a" title="Permalink to this headline">¶</a></h4>
<p>Specifies the reference genome of the GWAS summary statistics data. If left off when a GWAS file is uploaded, the reference genome is assumed to be the same as the samples reference genome (-r).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>-a hg19
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="calculate-scores">
<h2>Calculate Scores<a class="headerlink" href="#calculate-scores" title="Permalink to this headline">¶</a></h2>
<p>Polygenic risk scores can be calculated directly through the command-line or through the interactive menu. Using just the required parameters, the CLI will calculate risk scores for all studies in the database for each individual in the input file. Additional parameters will filter studies to be included in the calculation.</p>
</div>
</div>
<div class="section" id="uploading-gwas-summary-statistics">
<h1>Uploading GWAS Summary Statistics<a class="headerlink" href="#uploading-gwas-summary-statistics" title="Permalink to this headline">¶</a></h1>
<p>In addition to calculating polygenic risk scores using GWA studies from the GWAS Catalog stored in our database, users have the option to upload their own GWAS summary statistics to use in risk score calculations.</p>
<div class="section" id="format">
<h2>Format<a class="headerlink" href="#format" title="Permalink to this headline">¶</a></h2>
<p>The GWAS summary statistics file to be uploaded <strong>must</strong> be in the correct format. It should be either a .tsv or a .txt tab separated file, or a zipped .tsv or .txt. The following columns are required and must be included in the file’s header line: Study ID, Trait, Rsid, Chromosome, Position, Risk Allele, Odds Ratio, and P-value. Additional optional columns that will be included if present are: Citation and Reported Trait. Column order does not matter and there may be extra columns present in the file. Required and optional header names must be exact.</p>
<p>If more than one odds ratio exists for an Rsid in a study, the odds ratio and corresponding risk allele with the most significant p-value will be used. Additonally, though we perform strand flipping on GWAS summary statistics data we use from the GWAS Catalog, we do not perform strand flipping on uploaded data. Please ensure that your data is presented on the correct strand.</p>
<p><em>NOTE: If a GWAS data file is specified, risk scores will only be calculated on that data. No association data from the PRSKB will be used. Additionally, the optional params -t, -k, -i, -e, and -g will be ignored.</em></p>
</div>
<div class="section" id="columns">
<h2>Columns<a class="headerlink" href="#columns" title="Permalink to this headline">¶</a></h2>
<p>Below is a brief overview of the required and optional columns for uploading GWAS summary statistics data.</p>
<div class="section" id="required-columns">
<h3>Required Columns<a class="headerlink" href="#required-columns" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li><p><strong>Study ID</strong> - A unique study identifyer. In our database, we use GWAS Catalog study identifiers. As long as this is unique for each study, it can be whatever you want.</p></li>
<li><p><strong>Trait</strong> - The Experimental Factor Ontology (EFO) trait the GWAS deals with.</p></li>
<li><p><strong>Rsid</strong> - The Reference SNP cluster ID (Rsid) of the SNP.</p></li>
<li><p><strong>Chromosome</strong> - The chromosome the SNP resides on.</p></li>
<li><p><strong>Position</strong> - The position of the SNP in the reference genome.</p></li>
<li><p><strong>Risk Allele</strong> - The allele that confers risk or protection.</p></li>
<li><p><strong>Odds Ratio</strong> - Computed in the GWAS study, a numerical value of the odds that those in the case group have the allele of interest over the odds that those in the control group have the allele of interest.</p></li>
<li><p><strong>P-value</strong> - The probability that the risk allele confers the amount of risk stated.</p></li>
</ol>
</div>
<div class="section" id="optional-columns">
<h3>Optional Columns<a class="headerlink" href="#optional-columns" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li><p><strong>Citation</strong> - The citation information for the study.</p></li>
<li><p><strong>Reported Trait</strong> - Trait description for this study in the authors own words.</p></li>
</ol>
</div>
</div>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>Run the calculator on all studies in the database:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS
</pre></div>
</div>
<p>Run the calculator on all studies about the trait ‘acne’:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS -t acne
</pre></div>
</div>
<p>Run the calculator on all high impact studies with samples that are of European descent:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p AMR -k HI -e european
</pre></div>
</div>
<p>Run the calculator on the study with study ID GCST004410:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p EUR -i GCST004410
</pre></div>
</div>
<p>Run the calculator on all studies in the database in two steps:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS -s <span class="m">1</span>
./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS -s <span class="m">2</span>
</pre></div>
</div>
<p>Run the calculator on all studies about the trait ‘acne’, filtering studies from the allAssociations_hg19_f.txt working file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS
./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS -t acne -s <span class="m">2</span>
</pre></div>
</div>
<p>Run the calculator using uploaded GWAS summary statistics:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p AFR -u path/to/GWAS/GWASsummaryStatistics.tsv -a hg17
./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p AFR -u path/to/GWAS/GWASsummaryStatistics.tsv
</pre></div>
</div>
<p>More examples can be found in the CLI download README.md</p>
</div>
<div class="section" id="output-results">
<h1>Output Results<a class="headerlink" href="#output-results" title="Permalink to this headline">¶</a></h1>
<p>There are two choices for the tsv output results - condensed (default) or full.</p>
<div class="section" id="condensed">
<h2>Condensed<a class="headerlink" href="#condensed" title="Permalink to this headline">¶</a></h2>
<p>This version of the output results contains one row for each study with columns for each sample’s polygenic risk score.</p>
<p>Study ID | Citation | Reported Trait | Trait(s) | Sample1 | Sample2 | Sample3 | ect.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS
</pre></div>
</div>
</div>
<div class="section" id="full">
<h2>Full<a class="headerlink" href="#full" title="Permalink to this headline">¶</a></h2>
<p>This version of the output results contains one row for each sample/study pair. It also includes columns listing the rsIDs of the snps involved in the risk score calculation.</p>
<p>Sample | Study ID | Citation | Reported Trait | Traits(s) | Risk Score | Protective Variants | Risk Variants | Variants Without Risk Allele | Variants in High LD</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./runPrsCLI.sh -f path/to/file/samples.vcf -o path/to/file/output.tsv -c <span class="m">0</span>.0005 -r hg19 -p SAS -v
</pre></div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">Polygenic Risk Score Knowledge Base</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Madeline Page, Elizabeth Ward, Matthew Cloward, Ed Ringger, Louisa Dayton.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.5.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>